### PR TITLE
[herd] Variant -self : updates for running cat files

### PR DESCRIPTION
**Pull request by artkhyzha**
This PR makes several basic changes:
* Added IC IALLU
* Added Fetch actions as a subcategory of memory accesses
* Added NoAction actions in order to display an event for the non-fetch part B/NOP,
* Added (drafts of) adaptation of cos.cat

One aspect this PR falls short in is the display of herd's output: currently ifetch-cos.cat redefines rf and po in order to exclude some edges from those, but the excluded edges are still visible in the output (although denoted differently).

### CODE REVIEW

**Comment by maranget** [comment]
> One aspect this PR falls short in is the display of herd's output: currently ifetch-cos.cat redefines rf and po in order to exclude some edges from those, but the excluded edges are still visible in the output (although denoted differently).

Hi @artkhyzha, for some reason the `unshow` directives in your file `ifetch-cos.cat` do not work on `po` and `rf`. However, the `unshow` directive in a configuration file or the `-unshow` command line option do work. I attach a configuration file that you can use to get the desired effect of not showing `rf` and `po`. Notice that not showing `po`  hampers diagrams legibility IMHO. The configuration file [self.txt](https://github.com/herd/herdtools7/files/7788573/self.txt) (to be used as argument to command-line option `-conf`) performs additional configurations such as color changes.

File: herd/memUtils.ml [hunk 1]
```diff
@@ -488,14 +488,14 @@ let lift_proc_info i evts =
...
-      | Dir.R,Dir.W ->
+      | Dir.R,Dir.W| Dir.F,Dir.W ->
           begin match
             find_source rfmap e1
           with
           | S.Store w1 -> E.EventRel.add (w1,e2) k
           | S.Init -> k
           end
-      | Dir.W,Dir.R ->
+      | Dir.W,Dir.R| Dir.W,Dir.F ->
```
Comment (maranget) [inline herd/memUtils.ml:498]:
```suggestion
      | Dir.W,(Dir.R|Dir.F) ->
```

**Review by maranget — COMMENTED** [review]
Hi @artkhyzha, here a few comments, most of them on using or-patterns more. I have not looked at the `ifetch-cos.cat` file

File: herd/machAction.ml [hunk 2]
```diff
@@ -302,6 +322,7 @@ end = struct
 
   let is_i = function
     | DC(op,_) as a -> is_dc a && AArch64Base.DC.i op
+    | IC _ -> Warn.warn_always "FIXME possibly mistaking IC for DC" ; true
```
[the remark this answers is not in the document]
Comment (artkhyzha) [reply herd/machAction.ml:324]: Yes, could be deleted. That said, should `is_i` even be modified at all? (My worry was that `is_i` is used somewhere in a way that is not prepared for a new cache maintenance instruction. Can't find traces of that though, and IC should not have previously appeared in AArch64 cat files anyway.)

File: herd/AArch64Sem.ml [hunk 3]
```diff
@@ -1384,6 +1389,27 @@ module Make
...
+          M.mk_singleton_es (Act.IC (op, None)) ii >>! B.Next
+        else 
+        begin (* IC IVAU *)
+          let ic_loc op a ii =
+            let mk_act loc = Act.IC (op,Some loc) in
+            let loc = A.Location_global a in
+            M.mk_singleton_es (mk_act loc) ii in
+          let mop _ac a = ic_loc op a ii in
+          let dir = Dir.W in
+          lift_memop dir false
```
[the remark this answers is not in the document]
Comment (artkhyzha) [reply herd/AArch64Sem.ml:1403]: Good point, this will get simplified away in the next commit.

**Comment by maranget** [comment]
Hi @artkhyzha, there is a conceptual difficulty: one have to transmit the link register update from the branch and link instruction to the return instruction. This means adding a specific constructor to the Branch.t type. I'll send you a tanative patch soon.

**Comment by maranget** [comment]
@artkhyzha Thanks a lot merged.

**Pull request merged**